stdlib/public/core/Set.swift

```diff
@@ -2013,259 +2016,268 @@ extension _NativeSet/*: _SetBuffer*/ where Element: Hashable {
-    return k._rawHashValue(seed: _storage.seed) & _bucketMask
-  }
-
-  @inlinable // FIXME(sil-serialize-all)
-  internal func _index(after bucket: Int) -> Int {
-    // Bucket is within 0 and bucketCount. Therefore adding 1 does not overflow.
-    return (bucket &+ 1) & _bucketMask
-  }
-
-  @inlinable // FIXME(sil-serialize-all)
-  internal func _prev(_ bucket: Int) -> Int {
-    // Bucket is not negative. Therefore subtracting 1 does not overflow.
-    return (bucket &- 1) & _bucketMask
+extension _NativeSet {
+  @inlinable
+  @inline(__always)
+  internal func hashValue(for element: Element) -> Int {
+    return element._rawHashValue(seed: _storage.seed)
   }
 
-  @inlinable // FIXME(sil-serialize-all)
+  @inlinable
   @inline(__always)
-  internal func _find(_ key: Element) -> (pos: Index, found: Bool) {
-    return _find(key, startBucket: _bucket(key))
+  internal func find(_ element: Element) -> (index: Index, found: Bool) {
+    return find(element, hashValue: self.hashValue(for: element))
   }
 
   /// Search for a given key starting from the specified bucket.
   ///
   /// If the key is not present, returns the position where it could be
   /// inserted.
-  @inlinable // FIXME(sil-serialize-all)
+  @inlinable
   @inline(__always)
-  internal func _find(_ key: Element, startBucket: Int)
-    -> (pos: Index, found: Bool) {
-
-    var bucket = startBucket
-
-    // The invariant guarantees there's always a hole, so we just loop
-    // until we find one
-    while true {
-      let isHole = !isInitializedEntry(at: bucket)
-      if isHole {
-        return (Index(bucket: bucket), false)
-      }
-      if self.key(at: bucket) == key {
-        return (Index(bucket: bucket), true)
+  internal func find(
+    _ element: Element,
+    hashValue: Int
+  ) -> (index: Index, found: Bool) {
+    var (index, found) = _storage.hashTable.lookupFirst(hashValue: hashValue)
+    while found {
+      if uncheckedElement(at: index) == element {
+        return (index, true)
       }
-      bucket = _index(after: bucket)
+      (index, found) =
+        _storage.hashTable.lookupNext(hashValue: hashValue, after: index)
     }
+    return (index, false)
   }
+}
 
-  @usableFromInline @_transparent
-  internal static func bucketCount(
-    forCapacity capacity: Int,
-    maxLoadFactorInverse: Double
-  ) -> Int {
-    // `capacity + 1` below ensures that we don't fill in the last hole
-    return Swift.max(
-      Int((Double(capacity) * maxLoadFactorInverse).rounded(.up)),
-      capacity + 1)
+// This function has a highly visible name to make it stand out in stack traces.
+@usableFromInline
+@inline(never)
+internal func ELEMENT_TYPE_OF_SET_VIOLATES_HASHABLE_REQUIREMENTS(
+  _ elementType: Any.Type
+) -> Never {
+  fatalError("""
+    Duplicate elements of type '\(elementType)' were found in a Set.
+    This usually means that either the type violates Hashable's requirements, or
+    that members of such a set were mutated after insertion.
```

Comment:
the case we have seen often in Foundation is where an element of a NSMutableSet has a non-stable hash value. is this function breakpointable to help developers figure out why they are getting this failure?

Reply:
While the function is declared internal, the `@usableFromInline` attribute means there will still be a symbol generated for it -- so it'll be breakpointable. (It essentially makes the symbol public on the ABI level, so lldb will see it.)

I'll have to play with this a little to make sure the idea is sound. Multiline error messages may cause issues. Also, I think `fatalError` generates a bunch of stack frames before aborting, so I'm not sure if this will be as visible as I hope. Plus I think it may be useful to print one of the duplicate values (passed in as an `Any` parameter), along with the type -- although that may be a step too far.